# NumPy Basics

why use NumPy ?

NumPy is a powerful library for numerical computing in Python. Here are some reasons why you might want to use NumPy:

1. NumPy provides efficient storage
2. It also provides better ways of handling data(functions) for processing
3. It is fast
4. It is easy to learn
5. NumPy uses relatively less memory to store data


In [1]:
import numpy as np

np array creation methods

1. Conversion from other Python source(tuple,list)
2. Intrinsic numpy array creation object(arange, ones, zeros)
3. Reading arrays from disk, either from standard or custom formats
4. Creating arrays from raw bytes through use of strings or buffers
5. Use of special library function ex - random
6. Replicating, joining, or mutating existing arrays

### Conversion from other Python source(tuple,list)

In [2]:
# from int
myarr = np.array([[3,6,32,7]],np.int32)

In [3]:
myarr[0,1].item()     #item function will return literal value

6

if we give myarr[1,1].item() -> IndexError: index 1 is out of bounds for axis 0 with size 1

creating an 1d numpy array

In [4]:
myarr1 = np.array([3,6,32,7],np.int32)

In [5]:
myarr1[1].item()

6

In [6]:
myarr.shape     # returns array shape(matrix) : 1,4 means 1 row and 4 columns

(1, 4)

In [7]:
myarr.size # length

4

In [8]:
myarr.dtype     # give type of data

dtype('int32')

In [9]:
# from tuple

arr_tuple = np.array({1,2,3,4})
arr_tuple


array({1, 2, 3, 4}, dtype=object)

### from intrinsic np array creation arange,ones,zeros

In [10]:
zeros = np.zeros((2,5))

In [11]:
zeros # returns 2,5 arrays

array([[0., 0., 0., 0., 0.],
       [0., 0., 0., 0., 0.]])

In [12]:
zeros.dtype

dtype('float64')

In [13]:
zeros.shape

(2, 5)

In [14]:
rng = np.arange(15)

In [15]:
rng # gives a array from 0 to n-1

array([ 0,  1,  2,  3,  4,  5,  6,  7,  8,  9, 10, 11, 12, 13, 14])

In [16]:
lspace = np.linspace(1,50,10) # returns equally linerally spaced from 1 to 12 

In [17]:
lspace

array([ 1.        ,  6.44444444, 11.88888889, 17.33333333, 22.77777778,
       28.22222222, 33.66666667, 39.11111111, 44.55555556, 50.        ])

In [18]:
emp = np.empty((4,6))

In [19]:
emp

array([[0., 0., 0., 0., 0., 0.],
       [0., 0., 0., 0., 0., 0.],
       [0., 0., 0., 0., 0., 0.],
       [0., 0., 0., 0., 0., 0.]])

In [20]:
emp_like = np.empty_like(lspace) # takes a old array size and create a new array

In [21]:
emp_like

array([ 1.        ,  6.44444444, 11.88888889, 17.33333333, 22.77777778,
       28.22222222, 33.66666667, 39.11111111, 44.55555556, 50.        ])

In [22]:
ide = np.identity(45)

In [23]:
ide

array([[1., 0., 0., ..., 0., 0., 0.],
       [0., 1., 0., ..., 0., 0., 0.],
       [0., 0., 1., ..., 0., 0., 0.],
       ...,
       [0., 0., 0., ..., 1., 0., 0.],
       [0., 0., 0., ..., 0., 1., 0.],
       [0., 0., 0., ..., 0., 0., 1.]])

In [24]:
ide.shape

(45, 45)

In [25]:
arr = np.arange(99)

In [26]:
arr

array([ 0,  1,  2,  3,  4,  5,  6,  7,  8,  9, 10, 11, 12, 13, 14, 15, 16,
       17, 18, 19, 20, 21, 22, 23, 24, 25, 26, 27, 28, 29, 30, 31, 32, 33,
       34, 35, 36, 37, 38, 39, 40, 41, 42, 43, 44, 45, 46, 47, 48, 49, 50,
       51, 52, 53, 54, 55, 56, 57, 58, 59, 60, 61, 62, 63, 64, 65, 66, 67,
       68, 69, 70, 71, 72, 73, 74, 75, 76, 77, 78, 79, 80, 81, 82, 83, 84,
       85, 86, 87, 88, 89, 90, 91, 92, 93, 94, 95, 96, 97, 98])

In [27]:
arr = arr.reshape(3,33)   # reshapes the array use only factors

In [28]:
arr

array([[ 0,  1,  2,  3,  4,  5,  6,  7,  8,  9, 10, 11, 12, 13, 14, 15,
        16, 17, 18, 19, 20, 21, 22, 23, 24, 25, 26, 27, 28, 29, 30, 31,
        32],
       [33, 34, 35, 36, 37, 38, 39, 40, 41, 42, 43, 44, 45, 46, 47, 48,
        49, 50, 51, 52, 53, 54, 55, 56, 57, 58, 59, 60, 61, 62, 63, 64,
        65],
       [66, 67, 68, 69, 70, 71, 72, 73, 74, 75, 76, 77, 78, 79, 80, 81,
        82, 83, 84, 85, 86, 87, 88, 89, 90, 91, 92, 93, 94, 95, 96, 97,
        98]])

In [29]:
arr = arr.ravel()   # returns a 1D array
arr

array([ 0,  1,  2,  3,  4,  5,  6,  7,  8,  9, 10, 11, 12, 13, 14, 15, 16,
       17, 18, 19, 20, 21, 22, 23, 24, 25, 26, 27, 28, 29, 30, 31, 32, 33,
       34, 35, 36, 37, 38, 39, 40, 41, 42, 43, 44, 45, 46, 47, 48, 49, 50,
       51, 52, 53, 54, 55, 56, 57, 58, 59, 60, 61, 62, 63, 64, 65, 66, 67,
       68, 69, 70, 71, 72, 73, 74, 75, 76, 77, 78, 79, 80, 81, 82, 83, 84,
       85, 86, 87, 88, 89, 90, 91, 92, 93, 94, 95, 96, 97, 98])

In [30]:
arr.shape

(99,)

Axis in numpy

in 1D , [1,2,3,4,5] -> total axis = 1 [Axis0]
in 2D, [[1,2,3]
        [4,5,6]
        [7,8,9]] -> total axis = 2 [Axis0,Axis1]

axis0 will be row(down) , axis1 will be column(right)
....................
in nD , [         ] -> total axis = n [Axis0, ... , Axis(n-1), Axisn]



In [31]:
x = [[1,2,3],[4,5,6],[7,1,0]]

In [32]:
ar = np.array(x)

In [33]:
ar

array([[1, 2, 3],
       [4, 5, 6],
       [7, 1, 0]])

In [34]:
ar.sum(axis=0)

array([12,  8,  9])

In [35]:
ar.sum(axis=1)

array([ 6, 15,  8])

In [36]:
arr_4d = np.ones((2,3,4,5))

In [37]:
arr_4d

array([[[[1., 1., 1., 1., 1.],
         [1., 1., 1., 1., 1.],
         [1., 1., 1., 1., 1.],
         [1., 1., 1., 1., 1.]],

        [[1., 1., 1., 1., 1.],
         [1., 1., 1., 1., 1.],
         [1., 1., 1., 1., 1.],
         [1., 1., 1., 1., 1.]],

        [[1., 1., 1., 1., 1.],
         [1., 1., 1., 1., 1.],
         [1., 1., 1., 1., 1.],
         [1., 1., 1., 1., 1.]]],


       [[[1., 1., 1., 1., 1.],
         [1., 1., 1., 1., 1.],
         [1., 1., 1., 1., 1.],
         [1., 1., 1., 1., 1.]],

        [[1., 1., 1., 1., 1.],
         [1., 1., 1., 1., 1.],
         [1., 1., 1., 1., 1.],
         [1., 1., 1., 1., 1.]],

        [[1., 1., 1., 1., 1.],
         [1., 1., 1., 1., 1.],
         [1., 1., 1., 1., 1.],
         [1., 1., 1., 1., 1.]]]])

In [38]:
arr_4d.shape

(2, 3, 4, 5)

In [39]:
arr_4d.sum(axis=0)

array([[[2., 2., 2., 2., 2.],
        [2., 2., 2., 2., 2.],
        [2., 2., 2., 2., 2.],
        [2., 2., 2., 2., 2.]],

       [[2., 2., 2., 2., 2.],
        [2., 2., 2., 2., 2.],
        [2., 2., 2., 2., 2.],
        [2., 2., 2., 2., 2.]],

       [[2., 2., 2., 2., 2.],
        [2., 2., 2., 2., 2.],
        [2., 2., 2., 2., 2.],
        [2., 2., 2., 2., 2.]]])

numpy attributes and methods

In [40]:
ar.flat

In [41]:
for i in ar.flat:
    print(i)

1
2
3
4
5
6
7
1
0


In [42]:
ar.ndim

2

In [43]:
ar.size    # no of elements in array

9

In [44]:
ar.nbytes    # total bytes consumed

72

In [45]:
one = np.array([1,3,4,643,2])

In [46]:
one.argmax().item()     # returns index of max element

3

In [47]:
one.argmin().item()     # returns index of min element

0

In [48]:
one.argsort()    # returns  array with indices if sorted

array([0, 4, 1, 2, 3])

In [49]:
ar

array([[1, 2, 3],
       [4, 5, 6],
       [7, 1, 0]])

In [50]:
ar.argmin().item() # first sort it ascending and then find min

8

In [51]:
ar.argmax().item()

6

In [52]:
ar.argmax(axis=0)     # gives an array of indices with maxvalue among the rows

array([2, 1, 1])

In [53]:
ar.argmax(axis=1)    # gives an array of indices with maxvalue among the columns

array([2, 2, 0])

In [54]:
ar.argsort(axis=1)

array([[0, 1, 2],
       [0, 1, 2],
       [2, 1, 0]])

In [55]:
ar.ravel()

array([1, 2, 3, 4, 5, 6, 7, 1, 0])

In [56]:
ar

array([[1, 2, 3],
       [4, 5, 6],
       [7, 1, 0]])

In [57]:
ar2 = np.array([[1, 2, 1],
       [4, 0, 6],
       [8, 1, 0]])

In [58]:
ar + ar2

array([[ 2,  4,  4],
       [ 8,  5, 12],
       [15,  2,  0]])

In [59]:
ar * ar2

array([[ 1,  4,  3],
       [16,  0, 36],
       [56,  1,  0]])

In [60]:
ar % ar2

/var/folders/dk/mlcg64md3_j_ggxlkxr4kxgr0000gn/T/ipykernel_1905/3841496730.py:1: RuntimeWarning: divide by zero encountered in remainder
  ar % ar2


array([[0, 0, 0],
       [0, 0, 0],
       [7, 0, 0]])

In [61]:
np.sqrt(ar)

array([[1.        , 1.41421356, 1.73205081],
       [2.        , 2.23606798, 2.44948974],
       [2.64575131, 1.        , 0.        ]])

In [62]:
ar.sum()

np.int64(29)

In [63]:
ar.max()

np.int64(7)

In [64]:
ar.min()

np.int64(0)

In [65]:
# find elements in an array in form of a tuple
np.where(ar>5)

(array([1, 2]), array([2, 0]))

In [66]:
np.count_nonzero(ar)   # no. of non zero elements

8

In [67]:
np.nonzero(ar)  #Returns a tuple of arrays, one for each dimension of a, containing the indices of the non-zero elements in that dimension. 

(array([0, 0, 0, 1, 1, 1, 2, 2]), array([0, 1, 2, 0, 1, 2, 0, 1]))

In [68]:
ar[1,2] = 0

In [69]:
np.nonzero(ar)

(array([0, 0, 0, 1, 1, 2, 2]), array([0, 1, 2, 0, 1, 0, 1]))

to verify how numpy uses less space

In [70]:
import sys

In [71]:
py_ar = [0,4,55,2]

In [72]:
np_ar = np.array(py_ar)

In [73]:
sys.getsizeof(1) * len(py_ar)

112

In [74]:
np_ar.itemsize * np_ar.size

32

np array taking only 32 bytes whereas python array took 112 bytes

In [75]:
np_ar.tolist()  # return python list of numpy array

[0, 4, 55, 2]